Add Egg Dropping Problem to Dynamic Programming folder

SidharthSarangi opened the pull request:
## Related Issue
-  Add Egg Dropping Puzzle problem issue  #215 

## Additional Info
Problem Statement:
You are given N identical eggs and you have access to a K-floored building from 1 to K.

There exists a floor f where 0 <= f <= K such that any egg dropped at a floor higher than f will break, and any egg dropped at or below floor f will not break. There are few rules given below. 

An egg that survives a fall can be used again.
A broken egg must be discarded.
The effect of a fall is the same for all eggs.
If the egg doesn't break at a certain floor, it will not break at any floor below.
If the eggs breaks at a certain floor, it will break at any floor above.
Return the minimum number of moves that you need to determine with certainty what the value of f is.

Link to the Problem [Link](https://www.geeksforgeeks.org/egg-dropping-puzzle-dp-11/)
#### Complexity Analysis: 
Time Complexity: O(n*k^2). 
Where 'n' is the number of eggs and 'k' is the number of floors, as we use a nested for loop 'k^2' times for each egg
Auxiliary Space: O(n*k). 
As a 2-D array of size 'n*k' is used for storing elements.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] ✅ My code follows the code style of this project.
- [x] 📝 My change requires a change to the documentation.
- [x] 🎀 I have updated the documentation accordingly.
- [x] 👀 I have read the [**CONTRIBUTING**](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [x] 🌟 ed the repo

 ##### Describe your changes in detail 
 I have added the fully working code for the problem egg dropping puzzle and changed the readme file accordingly and added all required information regarding the problem and explained there the working of the code with examples and provided important links.
<!--- Provide a general summary of your changes in the Title above -->

<!--- Why is this change required? What problem does it solve? -->

<!--- If it fixes an open issue, please link to the issue here. -->

<!--- Please describe in detail how you tested your changes. -->

<!--- Include details of your testing environment, and the tests you ran to -->

<!--- see how your change affects other areas of the code, etc. -->

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Output Screenshots
| Screenshot #1      |
| ----------- | ----------- |
| Full Code  | 
![Full code](https://user-images.githubusercontent.com/78564629/157181499-bb746047-cc0f-4ca9-b3ce-700269acac7e.png)
| Screenshot #2  |  
| Test Cases  |
![Egg](https://user-images.githubusercontent.com/78564629/157181709-65a02f7e-443b-4392-8591-7b0af18120d4.png)

SidharthSarangi commented:
Please review it @Lakhankumawat @iamgunjan @miraj0507 ,I have added the file and also updated the readme

Lakhankumawat commented:
Let me show you how to check the check list , there should be no space inside brackets like [x]
 `[x]`
- [x] Checklist
- [ ] Or you can just check them from outside without doing any of these
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [ ] 🌟 ed the repo

1. **Starr the repo if you haven't !**
2. Have you watched this [video](https://youtu.be/kIFwqp9B7pQ) ?


///----------------------PR README --------------------///


You need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms) 
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md) 
for more please go through these [here](https://youtu.be/7UgC3HDETRA)

SidharthSarangi commented:
Added the Problem statement and uploaded the screenshots as said.

SidharthSarangi commented:
starred the repo and uploaded the screenshots and also added the problem
statement with link. Also I am new to open source/ Github, so I am still
learning a lot of things and unaware of all functionality of github so
please help with the rest to get it merged. Looking forward to
contribute with many more solutions/code.

On Tue, Mar 8, 2022 at 8:57 AM LakhanKumawat ᵖ⁺ ***@***.***>
wrote:

> Let me show you how to check the check list , there should be no space
> inside brackets like [x]
> [x]
>
>    - Checklist
>    - Or you can just check them from outside without doing any of these
>    - ✨ I have added tests to cover my changes.
>    - 🚩 All new and existing tests passed.
>    - 🌟 ed the repo
>
>
>    1. *Starr the repo if you haven't !*
>    2. Have you watched this video <https://youtu.be/kIFwqp9B7pQ> ?
>
> ///----------------------PR README --------------------///
>
> You need to follow particular guidelines for your README , please follow
> these guidelines click here
> <https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms>
> also you can see how your README should look like see here
> <https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md>
> for more please go through these here <https://youtu.be/7UgC3HDETRA>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Lakhankumawat/LearnCPP/pull/310#issuecomment-1061368236>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASXM2FKE4VU7SDTXWOCQ5IDU63CIVANCNFSM5P3PEGQQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

SidharthSarangi commented:
Please check now @Lakhankumawat @iamgunjan

SidharthSarangi commented:
Please review it cause I want to work on other files/folder/topics and getting this merged will let me keep in mind what exactly I am supposed to do while creating a pull request. Thank You. @Lakhankumawat @iamgunjan

Lakhankumawat reviewed (changes requested):
Modify Readme and code implementation , change function and variable names atleast instead of copy pasting everything from gfg directly

Lakhankumawat reviewed (changes requested):
Also i don't think you followed README stylings in fact i mentioned earlier in comments 

You need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms)
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md)
for more please go through these [here](https://youtu.be/7UgC3HDETRA)

SidharthSarangi commented:
I have now updated the Readme the way you said following proper styling and the code/logic is completely mine except the driver code which is common.

Lakhankumawat reviewed (approved):
LGTM

SidharthSarangi commented:
Please review @iamgunjan

miraj0507 commented:
LGTM

SidharthSarangi commented:
Can someone mention why is it not getting merged? Even after I fixed everything I was required to do @Lakhankumawat @miraj0507

siddhi-244 commented:
> Can someone mention why is it not getting merged? Even after I fixed everything I was required to do @Lakhankumawat @miraj0507

Hey please resolve the merge conflicts!

n4i9kita commented on D-DynamicProgramming/EggDroppingPuzzle.cpp:
```diff
@@ -0,0 +1,55 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+class Solution
```
I think you should avoid making a Solution class in this problem. 
Is there any benefit here of doing so?

Also, I would recommend that you mention the header files that are being used explicitly, rather than `bits/stdc++.h`

SidharthSarangi replied:
#### Why we use class instead of function?
By using classes, you're ensuring that methods are only used on one set of data. This adds to the security of the code because you're less likely to use functions where they don't belong.

**P.S-** I get what you are trying to say but using class is a good practice when writing big programs.
             "bits/stdc++.h" I will take care of that from next time I am committing something.

n4i9kita reviewed (commented):
LGTM.

Good job.

SidharthSarangi commented:
> 



> > Can someone mention why is it not getting merged? Even after I fixed everything I was required to do @Lakhankumawat @miraj0507
> 
> Hey please resolve the merge conflicts!

I have resolved the merge conflicts please review!

Lakhankumawat reviewed (approved):
Your Contribution looks great 🌟🎉

And since it is your second PR and it is reviewed good by all mentors and you might also be frustrated by now let me increase the level of your contribution seeing your efforts 🎉

Lakhankumawat commented:
@all-contributors please add @SidharthSarangi  code ,
